### ISSUE TITLE
FSQ Oddness

### ISSUE THREAD

**Issue by zaptrem** [lucidrains/vector-quantize-pytorch#194]
VQPytorch's FSQ with symmetry on and noise dropping set to 0.5 seems to perform significantly better than the [reference implementation](https://github.com/Stability-AI/stable-codec/blob/main/stable_codec/fsq.py) in recon loss with the same settings, so I set out to figure out why suspecting one of the two impls may be broken. 

First
```py
    def symmetry_preserving_bound(self, z):
        """
        QL(x) = 2 / (L - 1) * [(L - 1) * (tanh(x) + 1) / 2 + 0.5] - 1
        """
        levels_minus_1 = (self._levels - 1)
        scale = 2.0 / levels_minus_1
        bracket = (levels_minus_1 * (torch.tanh(z) + 1) / 2.0) + 0.5
        return scale * bracket - 1.0
```

This simplifies to this.

```py
    def symmetry_preserving_bound(self, z):
        return torch.tanh(z) + 1.0 / (self._levels - 1)
```

Second, this version doesn't seem to match what they do in the reference implementation.

You do:
```py
symmetry_preserved_bound = torch.tanh(z) + 1.0 / (self._levels - 1)
rounded = round_ste(symmetry_preserved_bound) / half_width
```
s

They do (also simplified):
```
dfsq_scale_shift = (torch.tanh(z)/self.scale + 1) * (self._levels - 1) / 2
rounded = round_ste(dfsq_scale_shift)
dfsq_inverse_scale_shift = (rounded * self.scale * 2 / (self._levels - 1)) - self.scale
```

When you switch out the above part the performance difference occurs.

Third, the noise scaling is slightly different, idk how much this matters:

Yours:
```py
offset = (torch.rand_like(z) - 0.5) / (self._levels // 2)
quantized = torch.where(offset_mask, unquantized + offset, quantized)
```

Theirs
```py
offset = (torch.rand_like(z) - 0.5) * (self.scale * 2 / (self._levels - 1))
quantized = torch.where(mask, quantized, z + offset)
```

**Comment by zaptrem** [comment]
@lucasnewman Did you successfully train a working model with this FSQ bottleneck? I'm curious how much these differences negatively affect performance.

**Comment by lucasnewman** [comment]
No, I haven't tried anything outside of toy settings. This was originally based on the paper, not their reference implementation, fwiw — if you want to send a PR to align it with their implementation, that would be great!

**Comment by zaptrem** [comment]
> if you want to send a PR to align it with their implementation, that would be great!

I haven't been able to get it to work at all (including the unaltered one from their repo), so I don't trust the correctness of my changes enough to PR them.

**Comment by lucasnewman** [comment]
> > if you want to send a PR to align it with their implementation, that would be great!
> 
> I haven't been able to get it to work at all (including the unaltered one from their repo), so I don't trust the correctness of my changes enough to PR them.

I haven't tried their official repo, but that matches my experience with a small-scale ASR task -- regular FSQ worked better. Reproducing their entire neural audio codec training pipeline would be somewhat difficult, but if I find a few minutes I'll try the official one and see if it works any better for the ASR task.

**Comment by WJohnnyW** [comment]
@zaptrem @lucasnewman hi，I found that 
```
def symmetry_preserving_bound(self, z):
        levels_minus_1 = (self._levels - 1)
        scale = 2.0 / levels_minus_1
        bracket = (levels_minus_1 * (torch.tanh(z) + 1) / 2.0) + 0.5
        return scale * bracket - 1.0
```

already normalized by multipling `scale`, but when quantizing, it normalized again.

```
if preserve_symmetry:
        quantized = round_ste(self.symmetry_preserving_bound(z)) / half_width
```
I am not familiar with this field, wish I was wrong
